server/src/main/java/org/apache/druid/client/DataSourcesSnapshot.java

```diff
@@ -167,6 +326,112 @@ private List<DataSegment> determineOvershadowedSegments()
+    return ImmutableSet.copyOf(overshadowedSegments);
+  }
+
+  public static Set<SegmentWithOvershadowedStatus> getSegmentsWithOvershadowedStatus(
+      Collection<ImmutableDruidDataSource> segments,
+      Set<DataSegment> overshadowedSegments,
+      Map<String, Set<SegmentId>> handedOffState)
+  {
+
+    final Stream<DataSegment> usedSegments = segments
+        .stream()
+        .flatMap(t -> t.getSegments().stream());
+
+    return usedSegments
+        .map(segment -> new SegmentWithOvershadowedStatus(
+            segment,
+            overshadowedSegments.contains(segment),
+            getHandedOffStateForSegment(handedOffState, segment.getDataSource(), segment.getId())
+             )
+        )
+        .collect(Collectors.toSet());
+  }
+
+  private static boolean getHandedOffStateForSegment(
+      Map<String, Set<SegmentId>> handedOffState,
+      String dataSource, SegmentId segmentId
+  )
+  {
+    return handedOffState
+        .getOrDefault(dataSource, new HashSet<>())
+        .contains(segmentId);
+  }
+
+  private static CircularBuffer<ChangeRequestHistory.Holder<List<DataSegmentChange>>> computeChanges(
+      Set<SegmentWithOvershadowedStatus> oldSegments,
+      Set<SegmentWithOvershadowedStatus> currentSegments,
+      CircularBuffer<ChangeRequestHistory.Holder<List<DataSegmentChange>>> oldChanges
+  )
+  {
+    if (oldSegments.isEmpty()) {
+      return new CircularBuffer<>(CHANGES_QUEUE_MAX_SIZE);
+    }
+
+    // a segment is added to the change set, if following changes:
+    // segmentId
+    // overshadowed state
+    // handed off state
+
+    Map<SegmentId, SegmentWithOvershadowedStatus> oldSegmentsMap =
+        oldSegments
+            .stream()
+            .collect(Collectors.toMap(
+                segment -> segment.getDataSegment().getId(),
+                Function.identity()));
+
+    Map<SegmentId, SegmentWithOvershadowedStatus> currentSegmentsMap =
+        currentSegments
+            .stream()
+            .collect(Collectors.toMap(
+                segment -> segment.getDataSegment().getId(),
+                Function.identity()));
+
+    Set<SegmentWithOvershadowedStatus> segmentToBeRemoved = Sets.difference(oldSegments, currentSegments);
+    Set<SegmentWithOvershadowedStatus> segmentToBeAdded = Sets.difference(currentSegments, oldSegments);
+
+    List<DataSegmentChange> changeList = new ArrayList<>();
+
+    segmentToBeRemoved.forEach(segment -> {
+      SegmentWithOvershadowedStatus newSegment = currentSegmentsMap.get(segment.getDataSegment().getId());
+      if (null == newSegment) {
+        changeList.add(new DataSegmentChange(segment, false, DataSegmentChange.ChangeReason.SEGMENT_REMOVED));
+      }
+    });
+
+    segmentToBeAdded.forEach(segment -> {
+      SegmentWithOvershadowedStatus oldSegment = oldSegmentsMap.get(segment.getDataSegment().getId());
+      boolean handedOffStatusChanged = !Objects.equals(null != oldSegment && oldSegment.isHandedOff(), segment.isHandedOff());
```

Comment:
Avoid these repeated null checks to improve readability. You can instead return after line 410 and handle the rest of the cases separately. See following suggestion.